src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs

```diff
@@ -0,0 +1,409 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Formats.Asn1;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Security.Cryptography.X509Certificates
+{
+    /// <summary>
+    /// This class facilitates building a distinguished name for an X.509 certificate.
+    /// </summary>
+    /// <remarks>
+    /// When constructing the SEQUENCE OF Relative Distinguished Names, this builder
+    /// constructs the SEQUENCE OF in the opposite order which they were added to the
+    /// builder. For example:
+    /// <code>
+    /// builder.AddCommonName("Contoso");
+    /// builder.AddCountryOrRegion("US");
+    /// </code>
+    /// Will produce a SEQUENCE OF with the country or region first, and the common
+    /// name second. Because <see cref="X500DistinguishedName" /> follows RFC 4514
+    /// when converting a Distinguished Name to a string, it starts with the last
+    /// Relative Distinuished Name component, moving backward toward the first.
+    /// Because the builder creates the SEQUENCE OF in reverse, and
+    /// <see cref="X500DistinguishedName" /> is also in reverse, it gives the appearence
+    /// of the added attributes and the string representation of the Distinguished Name in
+    /// the same order.
+    /// </remarks>
+    public sealed class X500DistinguishedNameBuilder
+    {
+        private readonly List<byte[]> _encodedComponents = new List<byte[]>();
+        private readonly AsnWriter _writer = new AsnWriter(AsnEncodingRules.DER);
+
+        /// <summary>
+        /// Adds a Relative Distinguished Name attribute identified by an OID.
+        /// </summary>
+        /// <param name="oidValue">The OID of the attribute.</param>
+        /// <param name="value">The value of the attribute.</param>
+        /// <param name="stringEncodingType">
+        /// The encoding type to use when encoding the <paramref name="value" />
+        /// in to the attribute.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="oidValue" /> or <paramref name="value" /> is <see langword="null" />.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// <para>
+        ///   <paramref name="oidValue" /> is an empty string or not a valid OID.
+        /// </para>
+        /// <para>-or-</para>
+        /// <para>
+        ///   <paramref name="stringEncodingType" /> is not a type for character strings.
+        /// </para>
+        /// <para>-or-</para>
+        /// <para>
+        ///   <paramref name="value" /> is not encodable as defined by <paramref name="stringEncodingType" />.
+        /// </para>
+        /// </exception>
+        public void Add(string oidValue, string value!!, UniversalTagNumber? stringEncodingType = null)
+        {
+            ArgumentException.ThrowIfNullOrEmpty(oidValue);
```

Comment:
It was something I considered and ultimately decided the order in which parameters are checked for not-null wasn't something that strongly warranted changing anything. I did not write tests for this behavior though because I don't think it makes sense to write a test that locks us in to a specific throwing order for `ArgumentNullException`. The tests as written in such a way that each parameter is tested independently.